Reload the SegmentMetadata even for temporary forward index

somandal opened the pull request:
This PR fixes an issue where the `SegmentMetadata` isn't updated and reloaded for temporary forward index (which is created just to create other indexes off of and is then deleted once the other `IndexHandler` classes have run). The `SegmentMetadata` may be needed to construct the other indexes and should be updated.

cc @Jackie-Jiang @vvivekiyer @siddharthteotia

vvivekiyer commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseIndexHandler.java:
```diff
@@ -93,6 +100,24 @@ protected void createForwardIndexIfNeeded(SegmentDirectory.Writer segmentWriter,
     }
 
+    // Update the segmentMetadata
+    File indexDir = _segmentMetadata.getIndexDir();
```
As per my understanding, whenever we update metadata properties file, we also want to update the in-memory segmentMetadata data-structure. Is that correct?

Can we move the reloadSegmentMetadata() call closer to the place where the we actually update the file - or maybe wrap in-memory and file updates into a common function that can be called?

This will also make sure that future callers don't miss updating in-memory DS when updating the file.

somandal replied:
> As per my understanding, whenever we update metadata properties file, we also want to update the in-memory segmentMetadata data-structure. Is that correct?

Yes that's my understanding as well. 

> 

So one problem is that in the `BaseIndexHandler` I have the following code (unimportant code removed):

```
    invertedIndexAndDictionaryBasedForwardIndexCreator.regenerateForwardIndex();

...
...

    // Update the segmentMetadata
    File indexDir = _segmentMetadata.getIndexDir();
    reloadSegmentMetadata(indexDir, columnName);
```

`invertedIndexAndDictionaryBasedForwardIndexCreator.regenerateForwardIndex()` calls the metadata file update part. Even if I moved the in-memory update part closer to the file update in this scenario, I'll need to call `_segmentMetadata = new SegmentMetadataImpl(indexDir);` in this function since we're creating a new `SegmentMetadata` object, similar to the reason why I need to do this in `SegmentPreProcessor`. Is that okay?

vvivekiyer replied:
I was thinking we could just change the signature of `updateMetadataProperties` as follows:

`public static SegmentMetadata  updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String, String> metadataProperties, File indexDir)`

Now we can just call,
`_segmentMetadata = updateMetadataProperties(...) `everywhere. 

Does that work?

somandal replied:
That won't solve the issue where `_segmentMetadata` assigned inside the class won't reflect the new object outside the class :( 

The problem is even if I make your suggested change, I'll have to do the following:

```
  public static SegmentMetadata updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String,
      String> metadataProperties, File indexDir, SegmentDirectory segmentDirectory)
      throws IOException {
    PropertiesConfiguration propertiesConfiguration = SegmentMetadataUtils.getPropertiesConfiguration(segmentMetadata);
    for (Map.Entry<String, String> entry : metadataProperties.entrySet()) {
      propertiesConfiguration.setProperty(entry.getKey(), entry.getValue());
    }
    SegmentMetadataUtils.savePropertiesConfiguration(propertiesConfiguration);

    segmentMetadata = new SegmentMetadataImpl(indexDir);
    segmentDirectory.reloadMetadata();

    return segmentMetadata;
  }
```

Since a new object is created and returned and we assign that like this: `_segmentMetadata = updateMetadataProperties(...) `, it's still replacing the local class object and not the object outside the class.

vvivekiyer replied:
Yeah, checked your latest commit and it looks better now. Agreed that we'd still have to update the object outside the class to reflect the new metadata. Thanks for the making the change.

vvivekiyer commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/SegmentPreProcessor.java:
```diff
@@ -107,17 +107,20 @@ public void process()
+        // needed to construct other indexes like RangeIndex. This also needs to be done outside of the IndexHandler
+        // code since modifying the `_segmentMetadata` within the IndexHandler doesn't modify this object directly but
+        // creates a new one for use within the IndexHandler.
+        _segmentMetadata = new SegmentMetadataImpl(indexDir);
```
Based on my comment above -

if we wrap file and in-memory update into a common function and use it in all places where we want to update metadata properties, the call here is not needed. Is my understanding correct?

somandal replied:
We will still need the following part:

```
_segmentMetadata = new SegmentMetadataImpl(indexDir);
```

But can remove this part:

```
_segmentDirectory.reloadMetadata();
```

We need the first part because inside the `IndexHandler` since we're creating a new `SegmentMetadata` object, this object which was passed to the `IndexHandler` won't be updated but instead a new local `SegmentMetadata` object will be created. Without re-creating the `SegmentMetadata` object here, we'll be passing in the older `_segmentMetadata` without the latest changes made to the on-disk structure. Hope this makes sense. I can remove the `_segmentDirectory.reloadMetadata()` from here.

vvivekiyer replied:
Sounds good. Thanks!

vvivekiyer commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java:
```diff
@@ -218,20 +218,23 @@ public void regenerateForwardIndex()
+          e);
+    }
 
+    if (!_isTemporaryForwardIndex) {
```
Question:
Earlier, we weren't updating metadata properties for temporary forward index creations.

Now that we are updating properties, shouldn't there also be a cleanup step (where we revert these updates) when the temp forward index is removed? Is that already guaranteed? If yes, can you please add a comment stating how the cleanup is handled for temporary forward index?

vvivekiyer commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java:
```diff
@@ -218,20 +218,23 @@ public void regenerateForwardIndex()
+          e);
+    }
 
+    if (!_isTemporaryForwardIndex) {
```
Earlier, we weren't updating in-memory and file metadata properties for temporary forward index.

Now that we are changing this, can you please clarify (and also add a comment) as to how the metadata property updates would be reverted?

somandal replied:
So @Jackie-Jiang and I discussed this in the slack thread where this issue was brought up and we decided not to reset the metadata to the original one on deletion of the forward index. The main reason for this is that:

- Temporary forward index should only be created for scenarios where forward index is disabled, which means dictionary must be present. So metadata changes to toggle the `hasDictionary` flag will never happen. SV columns only modify dictionary related metadata when reconstructing the forward index.
- For MV columns we have two additional metadata values that get modified: `MAX_MULTI_VALUE_ELEMENTS` and `TOTAL_NUMBER_OF_ENTRIES`. These can change if the MV column has duplicates. We decided that changing these and leaving them behind in the metadata after deletion should be okay.

Does that make sense? Do you see any other reason to change the metadata back?

vvivekiyer replied:
This makes sense. You might want to add a comment stating this reasoning. It's not intuitive when reading code as to why there's no cleanup for temp index.

somandal replied:
done

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseIndexHandler.java:
```diff
@@ -41,14 +43,19 @@
+  // The segmentMetadata may need to be updated after creating the forward index
+  protected SegmentMetadata _segmentMetadata;
+
+  public BaseIndexHandler(SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig,
```
If we pass in `SegmentDirectory`, we don't need to pass in `SegmentMetadata` because it can be read from `SegmentDirectory`, and we don't want to keep 2 copies of it. When we need to reload the segment metadata, we can call `SegmentDirectory.reloadMetadata()`, then get the updated `SegmentMetadata` with `SegmentDirectory.getSegmentMetadata()`

somandal replied:
thanks for the suggestion, done!

Jackie-Jiang reviewed (commented):
Let's replace `SegmentMetadata` with `SegmentDirectory` in the constructor, and always read segment metadata from the segment directory. This way we can save one copy of the segment metadata, and always ensure it is consistent

vvivekiyer commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/utils/SegmentMetadataUtils.java:
```diff
@@ -50,11 +51,15 @@ public static void savePropertiesConfiguration(PropertiesConfiguration propertie
   }
 
-  public static void updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String, String> metadataProperties) {
+  public static SegmentMetadata updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String,
```
segmentMetadata is redundant here, right?  Can we can fetch it from segmentDirectory?

somandal replied:
ooh good point, thanks! fixed it

vvivekiyer reviewed (approved):
LGTM. Minor comment.

siddharthteotia commented:
Is it possible to add a test for the corner case scenario that @Jackie-Jiang  had brought up in offline discussion ?

somandal commented:
> Is it possible to add a test for the corner case scenario that @Jackie-Jiang had brought up in offline discussion ?

Added a test and fixed a where the local `columnMetadata` wasn't getting updated. Thanks for this suggestion!

somandal commented:
hey @Jackie-Jiang and @siddharthteotia can you review this? I've addressed all feedback so far

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseIndexHandler.java:
```diff
@@ -41,13 +41,17 @@
-  public BaseIndexHandler(SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig) {
-    _segmentMetadata = segmentMetadata;
+  // The segmentMetadata may need to be updated after creating the forward index
+  protected SegmentMetadata _segmentMetadata;
```
Since this can be modified during the index handling, should we not put it as a member variable, but always read it from the `_segmentDirectory`? The one inside `_segmentDirectory` should always be updated

somandal replied:
done

Jackie-Jiang reviewed (approved):
Thanks for cleaning it up
